### PR TITLE
OurDNA: Improve project table view and filtering

### PR DESCRIPTION
**Pull request by illusional**
OurDNA RFC: https://docs.google.com/document/d/1bG8oZQdhA3lEeqlvnCZ1n_P1OqaqDBPlMPB7oz275ns/edit#heading=h.sxmryphgwrnn

Aims:
- Improve the filtering and usability of the table grid on the project page
- Make this page generally more useful

Consider:
- How to hide / show columns for specific projects
- Exporting data to CSV / etc

Actual changes:

- [large changes]:
    - Split the project summary from participant grid
    - Further decompose component structure for participant grid
    - Use generic filters for participant grid instead of custom query format
    - Add ability to export changes
    - Added ability to write more expressive queries
    - Encode more params on the URL to make sharing better
    - Migrate in support for multiple external IDs for participants / samples.

- [unrelated changes]
    - Display improvements to ErrorBoundary
    - Upgrade Vite
    - Add startswith GenericFilter operator
    - Improve participant query, by allowing for more nested objects to query against
    - Add participant tests

### CODE REVIEW

File: web/src/pages/project/ProjectGrid.tsx [hunk 1]
```diff
@@ -1,441 +1,418 @@
...
+                                                ? '2px solid var(--color-border-color)'
+                                                : '1px solid var(--color-border-default)',
+                                        backgroundColor,
+                                    }}
+                                    key={`${s.id}sequencing_group.${field.name}`}
+                                    rowSpan={
+                                        (sg.assays ?? []).length > 0
+                                            ? (sg.assays ?? []).length
+                                            : undefined
+                                    }
```
Comment (illusional) [inline web/src/pages/project/ProjectGrid.tsx:375]:
```suggestion
                                    rowSpan={sg.assays?.length || undefined}
```

File: web/src/pages/project/ValueFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+export const ValueFilter: React.FC<IValueFilter> = ({
+    category,
+    position,
+    updateFilterValues,
+    ...props
+}) => {
+    // Use the combination of category and filterKey to turn that into the correct value
+    // Note filterValues is a ParticipantGridFilter object, which has nested keys for each
+    // category (except participant).
```
Comment (illusional) [inline web/src/pages/project/ValueFilter.tsx:28]:
```suggestion
    // category.
```

File: web/src/pages/project/ValueFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    // Nb: The server tells us the filterKey here, and for meta keys it is prefixed with 'meta.'
+    //  So check if the filterKey starts with 'meta.' to determine if it is a meta key, and
+    //  then check the [category].meta object for the value
+
+    if (!props.filterKey) return <>No key</>
+    const isMeta = props.filterKey?.startsWith('meta.')
+    // set name to the filterKey without the .meta prefix
+    const name = props.filterKey.replace(/^meta\./, '')
+
+    // if we are filtering on the participant level, check the filterValues directly
```
Comment (illusional) [inline web/src/pages/project/ValueFilter.tsx:39]:
```suggestion
```

File: web/src/pages/project/ValueFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    let optionsToCheck = props?.filterValues?.[category] || {}
+
+    if (isMeta) {
+        // get the meta bit from the filterValues
+        optionsToCheck = optionsToCheck?.meta || {}
+    }
+    const isHighlighted = !!optionsToCheck && name in optionsToCheck
+
+    // TODO: remove this type ignore
```
Comment (illusional) [inline web/src/pages/project/ValueFilter.tsx:48]:
```suggestion
```

**Comment by illusional** [comment]
Having some issues with the Form component in the ValueFilter, when it's wrapped in a Form, the sizing doesn't work correctly, but if it's removed I can't catch the "on enter" event.

image

**Review by dancoates — COMMENTED** [review]
Frontend is looking good! all comments are pretty small

File: web/src/pages/project/ValueFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                {/* } */}
+            </Input>
+        )
+    }
+    // debugger
+
+    return (
+        <Form onSubmit={onSubmit}>
+            <Form.Group inline style={{ padding: 0, margin: 0 }}>
+                <Form.Field style={{ padding: 0, margin: 0 }}></Form.Field>
```
Comment (dancoates) [inline web/src/pages/project/ValueFilter.tsx:216]: Is this empty field required here?

File: web/src/pages/project/ProjectGridContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                <Message.Header>Invalid page options</Message.Header>
+                <Message.List>
+                    {Object.entries(hasOptionErrors).map(([k, v]) => (
+                        <Message.Item key={k}>{v.error}</Message.Item>
+                    ))}
+                </Message.List>
+                {hasOptionErrors.filter && (
+                    <DictEditor
+                        jsonStr={hasOptionErrors.filter.value}
+                        onChange={(v) => setFilterValues(v, true)}
```
[the remark this answers is not in the document]
Comment (illusional) [reply web/src/pages/project/ProjectGridContainer.tsx:206]: Ah, a relic from a previous iteration, thank you :)

File: web/src/pages/project/DictEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        } catch (jsonErr) {
+            throw yamlErr
+        }
+    }
+}
+
+export const DictEditor: React.FunctionComponent<DictEditorProps> = ({
+    jsonStr,
+    yamlStr,
+    obj,
```
[the remark this answers is not in the document]
Comment (illusional) [reply web/src/pages/project/DictEditor.tsx:40]: Yeah awesome, totally agreed!

File: web/src/pages/project/ProjectGrid.tsx [hunk 8]
```diff
@@ -1,448 +1,224 @@
...
-                                    key={`filter-${name}-${i}`}
+                                <th
+                                    key={`${category}-${field.key}-${idx}`}
                                     style={{
                                         borderBottom: 'none',
-                                        borderLeft: first
-                                            ? '2px solid var(--color-border-color)'
-                                            : '1px solid var(--color-border-default)',
+                                        borderLeft:
+                                            idx === 0
```
[the remark this answers is not in the document]
Comment (illusional) [reply web/src/pages/project/ProjectGrid.tsx:163]: This is theoretically putting thicker borders on each section (family, participant, sample, ...)

File: web/src/pages/project/ValueFilter.tsx (hunk 5 above)
Comment (illusional) [reply web/src/pages/project/ValueFilter.tsx:216]: Nope, just bad copy paste from the original code

File: web/src/pages/project/ProjectGrid.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (dancoates) [reply web/src/pages/project/ProjectGrid.tsx]: Ah I see, that is a bit trickier to do then, happy to leave as is

File: db/python/tables/assay.py [hunk 9]
```diff
@@ -188,25 +195,90 @@ async def get_assay_type_numbers_by_batch_for_project(self, project: ProjectId):
...
+                    'type': 'a.type',
+                }
+            )
+            wheres.append(_filter)
+            values.update(_values)
+
+        _query = f"""
+            SELECT
+                a.id, a.sample_id, a.type, a.meta, s.project,
+                JSON_OBJECTAGG(ae.name, ae.external_id) as external_ids,
```
[the remark this answers is not in the document]
Comment (illusional) [reply db/python/tables/assay.py:259]: Ooh, can just delete all those functions!

File: db/python/tables/participant.py [hunk 10]
```diff
@@ -208,14 +401,20 @@ async def update_participant(
                 """
                 await self.connection.execute(
                     _audit_update_query,
-                    {'pid': participant_id, 'names': to_delete, 'audit_log_id': audit_log_id},
+                    {
+                        'pid': participant_id,
+                        'names': to_delete,
+                        'audit_log_id': audit_log_id,
+                    },
```
[the remark this answers is not in the document]
Comment (illusional) [reply db/python/tables/participant.py:408]: Yep, agreed. This is my VSCode settings being different, which is pretty antithetical to this, and some conflict between black / rust, I want to look into it after these big PRs go in (external IDs, participant grid, nested samples, and permission refresh)

File: models/base.py [hunk 11]
```diff
@@ -22,6 +24,20 @@ def from_dict(cls, d: dict):
...
+    if val is None:
+        return None
+    if isinstance(val, dict):
+        return val
+    if isinstance(val, bytes):
+        val = val.decode()
+    if isinstance(val, str):
+        return json.loads(val)
+
+    raise ValueError(f'Unknown type for meta: {type(val)}')
```
[the remark this answers is not in the document]
Comment (illusional) [reply models/base.py:38]: Oh yep, I forgot I was going to unify those in this PR. I've actioned this, combining them into models, as db depends on models, but models shouldn't depend on db.

File: models/models/participant.py [hunk 12]
```diff
@@ -67,8 +69,12 @@ def to_external(self):
             reported_gender=self.reported_gender,
             karyotype=self.karyotype,
             meta=self.meta,
-            samples=[s.to_external() for s in self.samples or []],
-            families=[f.to_external() for f in self.families or []],
+            samples=[
+                s.to_external() for s in self.samples  # pylint:disable=not-an-iterable
```
[the remark this answers is not in the document]
Comment (illusional) [reply models/models/participant.py:73]: It's not, I wanted to see if I could fix the type annotation with a default, but that proved challenging (as a default list means you use `Pydantic.Field(default_factory=list)`, but pylint specifically did not like this and but was not reverted correctly.

File: test/test_web.py [hunk 13]
```diff
@@ -164,28 +312,57 @@ def get_test_participant_2():
...
+    #     ('id', 'Sequencing Group ID'),
+    #     ('platform', 'Platform'),
+    #     ('technology', 'Technology'),
+    #     ('type', 'Type'),
+    # ],
+    # assay_keys=[
+    #     ('type', 'type'),
+    #     ('meta.batch', 'batch'),
+    #     ('meta.reads_type', 'reads_type'),
+    # ],
```
[the remark this answers is not in the document]
Comment (illusional) [reply test/test_web.py:363]: Old style, was useful for reference but tests pass so it should've been removed.

File: db/python/filters/generic.py [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+            eq=func(self.eq) if self.eq else None,
+            in_=list(map(func, self.in_)) if self.in_ else None,
+            nin=list(map(func, self.nin)) if self.nin else None,
+            gt=func(self.gt) if self.gt else None,
+            gte=func(self.gte) if self.gte else None,
+            lt=func(self.lt) if self.lt else None,
+            lte=func(self.lte) if self.lte else None,
+            contains=func(self.contains) if self.contains else None,
+            icontains=func(self.icontains) if self.icontains else None,
+        )
```
[the remark this answers is not in the document]
Comment (illusional) [reply db/python/filters/generic.py]: Yes, nice catch!

File: db/python/tables/participant.py [hunk 15]
```diff
@@ -586,7 +587,11 @@ async def get_external_ids_by_participant(
         GROUP BY participant_id
         """
         rows = await self.connection.fetch_all(_query, {'pids': participant_ids})
-        return {r['id']: from_db_json(r['external_ids']) for r in rows}
+        retval: dict[int, list[str]] = {}
+        for row in rows:
+            external_ids = parse_sql_dict(row['external_ids'])
```
Comment (jmarshall) [inline db/python/tables/participant.py:592]: No, this SQL uses JSON_ARRAYAGG not JSON_OBJECTAGG. I guess a useful clarification would be to change the field name to `… AS external_ids_list` in the SQL to signpost this difference from other queries.

File: db/python/tables/participant.py [hunk 16]
```diff
@@ -581,17 +581,13 @@ async def get_external_ids_by_participant(
...
         WHERE participant_id IN :pids
         GROUP BY participant_id
         """
         rows = await self.connection.fetch_all(_query, {'pids': participant_ids})
-        retval: dict[int, list[str]] = {}
-        for row in rows:
-            external_ids = parse_sql_dict(row['external_ids'])
-            retval[row['id']] = list(external_ids.values()) if external_ids else []
-        return retval
+        return {r['id']: json.load(r['external_ids_list']) for r in rows}
```
Comment (jmarshall) [inline db/python/tables/participant.py:590]:
```suggestion
        return {r['id']: json.loads(r['external_ids_list']) for r in rows}
```

Of course, the real problem is that `get_external_ids_by_participant()` doesn't have a test case, so the unit tests aren't telling you about this :smile:

Comment (illusional) [reply db/python/tables/participant.py:590]: I should stop committing code on the train

Comment (jmarshall) [reply db/python/tables/participant.py]: :rofl:

Comment (illusional) [reply db/python/tables/participant.py]: Ooh funnily enough there is a test that covers it: `test.test_search.TestSample.test_search_mixed` which just failed, at least for my latest commit

**Review by jmarshall — APPROVED** [review]

**Pull request merged**